File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int result = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int res = splitPathSum(root);
        return Math.max(res, result);
    }
    private int splitPathSum(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int l = Math.max(splitPathSum(root.left), 0);
        int r = Math.max(splitPathSum(root.right), 0);
        result = Math.max(result, l + r + root.val);
        return Math.max(l, r) + root.val;
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };

 */

class Solution {
public:
    void inorder(TreeNode* root, int& cnt, int& res) {
        if (!root) return;
        if (cnt < 0) return;
        inorder(root->left, cnt, res);
        cnt--;
        if (cnt == 0) {
            res = root->val;
        }
        inorder(root->right, cnt, res);
    }
    int kthSmallest(TreeNode* root, int k) {
        int count = k, result = INT_MIN;
        inorder(root, count, result);
        return result;
    }
};
